---
title: "golub"
author: "Cindy Wang"
date: "9/20/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## install package

```{r, eval=FALSE}
source("https://bioconductor.org/biocLite.R")
biocLite("golubEsets")
```

```{r, message=FALSE}

library(golubEsets)
library(tidyverse)
library(DT)
library(caret)

source("heatmap.3.R")
```
## Data

- Description

The data are from Golub et al. These are the training samples, 27 patients with acute lymphoblastic
leukemia (ALL) and 11 patients with acute myeloid leukemia (AML). The samples were assayed
using Affymetrix Hgu6800 chips and data on the expression of 7129 genes (Affymetrix probes) are
available. The data were obtained from the Web site listed below and transformed slightly. They
were installed in an exprSet.

- Source

http://www-genome.wi.mit.edu/mpr/data_set_ALL_AML.html, after some anonymous Bioconductor
massaging

## Expression data distrubution

```{r}
data("Golub_Merge")
expr <- exprs(Golub_Merge)

# The first 59 rows are controls. Remove.

expr <- expr[60:nrow(expr), ]
expr.long <- as.data.frame(expr) %>%
  rownames_to_column(var="probeset") %>%
  gather(Sample, Expr, -probeset)

g <- ggplot(expr.long, aes(x=Expr, group=Sample)) +
  stat_density(aes(color = Sample), geom="line", position="identity") +
  theme(legend.position="none")

g


```
The data was obviously centered and the values seems to linear to original intensity values

A zoom-in view

```{r, message=F}
g + xlim(c(-1000, 1000))
```


### Data from website

```{r}
dta.web.train <- read.table("data_set_ALL_AML_train.txt", sep="\t", header=T, check.names=F, quote="", as.is=T, comment.char="", na.strings="", row.names=NULL)

row.ctls <- grep(" control)", dta.web.train$`Gene Description`)

dta.train <- dta.web.train[-row.ctls, seq(3, ncol(dta.web.train), 2)]
rownames(dta.train) <- dta.web.train$`Gene Accession Number`[-row.ctls]

call.train <- dta.web.train[-row.ctls, seq(4, ncol(dta.web.train), 2)]
colnames(call.train) <- colnames(dta.train)
rownames(call.train) <- rownames(dta.train)

```
 Test data
 
```{r}
dta.web.test <- read.table("data_set_ALL_AML_independent.txt", sep="\t", header=T, check.names=F, quote="", as.is=T, comment.char="", na.strings="", row.names=NULL)

row.ctls <- grep(" control)", dta.web.train$`Gene Description`)

dta.test <- dta.web.test[-row.ctls, seq(3, ncol(dta.web.test), 2)]
rownames(dta.test) <- dta.web.test$`Gene Accession Number`[-row.ctls]

call.test <- dta.web.test[-row.ctls, seq(4, ncol(dta.web.test), 2)]
colnames(call.test) <- colnames(dta.test)
rownames(call.test) <- rownames(dta.test)

```
```{r}
dta.train.long <- dta.train %>%
  rownames_to_column(var="probeset") %>%
  gather(Sample, Expr, -probeset)

g <- ggplot(dta.train.long, aes(x=Expr, group=Sample)) +
  stat_density(aes(color = Sample), geom="line", position="identity")

g
```

This looks like the same data set in the golubEset package

```{r}
all(expr[, colnames(dta.train)] == dta.train)
```

Summarise present calls

```{r}
call.train.long <- call.train %>%
  rownames_to_column(var="Probeset") %>%
  gather(Sample, Call, -Probeset) %>%
  rowwise() %>%
  mutate(Sample = strsplit(Sample))

g <- ggplot(call.train.long, aes(x=factor(as.numeric(Sample)))) + 
  geom_bar(aes(fill=Call)) +
  theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=0),
        axis.title.x = element_text("Sample"))
g
```

Roughly 1/3 of the probesets get call as P


The data profile does not looks like raw data but an output from some format of quantile normalization or similar methods. 

In order to use some commonly used tools for microarray data analysis, the data has to follow an approximate normal distribution. unfortunately, simple log2 transformation does not work because the data does not follow a typical un-log raw data that was typical for microarray data. It's very likely the data was centerd with some filtering and model fitting done, which was a common practice for microarray data. It's impossible to reverse engineer to convert back to orginal form however. Under this circumstance, one possible solution is to rank all genes and use ranking for downstream analysis. The drawback is this transformation results in linear a linear profile which deviate too far from natrual microarray data distribution. Another option is to use 1 to surrogate all negative values and do a log2 transformation. Considered about 2/3 of probesets got Absence calls, it is likely probesets with negative values can be confidently considered as absent. Therefore, surrogate all negative values with 1 is not likely to create significant artifect in downstream analysis.   

```{r}
expr.log <- log2(pmax(expr, 1))

expr.log.long <- as.data.frame(expr.norm) %>%
  rownames_to_column(var="Probeset") %>%
  gather(Sample, Expr, -Probeset)

ggplot(expr.log.long, aes(x=Expr, group=Sample)) +
  stat_density(aes(color = Sample), geom="line", position="identity")

```

Additional quantile normalization can be applied to align each samples. However, due to the data set was divided into training and testing, while quantile normalization is specific to the whole data set. Considered the overall expression profile align sufficiently well, we can skip quantile normalization.

```{r, eval=F}
require(preprocessCore)
expr.log.qnorm <- normalize.quantiles(expr.log)
rownames(expr.log.qnorm) <- rownames(expr.log)
expr.qnorm.long <- as.data.frame(expr.log.qnorm) %>%
  rownames_to_column(var="probeset") %>%
  gather(Sample, Expr, -probeset)

ggplot(expr.qnorm.long, aes(x=Expr, group=Sample)) +
  stat_density(aes(color = Sample), geom="line", position="identity")
```

Background correction may be helpful but CDF for this chip is not available anymore, there is no easy way to do it. 

Now we use Golub_Train data set to identify genes differentiate ALL and AML
```{r}
data("Golub_Train")
train <- exprs(Golub_Train)
# remove first 59 rows of controls
train <- train[60:nrow(train),]

train.log <- log2(pmax(train, 1))

train.pheno <- phenoData(Golub_Train)@data
```
```{r}
# density plot of transformed training data
train.log.long <- as.data.frame(train.log) %>%
  rownames_to_column(var="probeset") %>%
  gather(Sample, Expr, -probeset)

g <- ggplot(train.log.long, aes(x=Expr, group=Sample)) +
  stat_density(aes(color = Sample), geom="line", position="identity")
g

```

## limma 

limma is a commonly used method for differential expression analysis of microarray data. It is a modurated t-test aiming to combat low sample size issue of typical transcriptome expression data sets.

```{r}
require(limma)

# filter probeset with expr = 0 in more than 80% of samples
probesetFilter <- apply(train.log, 1, function(x) sum(x == 0 )/length(x) < 0.8)
train.log.ftd <- train.log[probesetFilter,]

# filter out probesets with mean log2Expr < 5
probesetFilter <- apply(train.log.ftd, 1, function(x) mean(x) >=3)
train.log.ftd <- train.log.ftd[probesetFilter,]

par(mfrow=c(7,6), mar=c(0, 0, 0, 0))
# for (i in 1:ncol(train.log.ftd)) {
for(i in c(1, 11, 21, 31)) {
  
  plotMA(train.log.ftd, array=i)
}
```

```{r}
# unloged MA plot
for (i in 1:ncol(train)) {
  plotMA(train, array=i)
}
```

```{r}
plotMA()
```

Setup experiment design as ALL and AML as two groups

```{r}
design <- train.pheno %>%
  dplyr::select(Samples, ALL.AML) %>%
  mutate(Value=1) %>%
  spread(ALL.AML, Value)

design[is.na(design)] <- 0

datatable(design)
```

```{r}
fit <- lmFit(train.log.ftd, design)
cont.matrix <- makeContrasts(AMLvsALL=AML-ALL, levels=design)
fit2 <- contrasts.fit(fit, cont.matrix)
fit2 <- eBayes(fit2)
tb <- topTable(fit2, adjust="BH", number=nrow(train.log.ftd)) %>%
  rownames_to_column(var="probeset")
```

```{r}
# function to check if the gene is in the 50 predictor genes in publication
inPub <- function(x) {
  if (x %in% ps.AML.hi) {
    label <- "AML_hi"
  } else if (x %in% ps.ALL.hi) {
    label <- "ALL_hi" 
  } else {
    label <- "None"
  }
  
  return(factor(label, levels=c("None", "AML_hi", "ALL_hi")))
}
```

```{r}
# volcano plot and highlight the genes in paper

ps.AML.hi <- c("M55150", "X95735", "U50136", "M16038", "U82759", "M23197", "M84526", "Y12670", "M27891", "X17042", "Y00787", "M96326", "U46751", "M80254", "L08246", "M62762", "M28130", "M63138", "M57710", "M69043", "M81695", "X85116", "M19045", "M83652", "X04085")
ps.ALL.hi <- c("U22376","X59417","U05259","M92287","M31211","X74262","D26156","S50223","M31523","L47738","U32944","Z15115","X15949","X63469","M91432","U29175","Z69881","U20998","D38073","U26266","M31303","Y08612","U35451","M29696","M13792")


tb <- tb %>%
  rowwise() %>%
  mutate(GeneSN = strsplit(probeset, "_")[[1]][1]) %>%
  mutate(publication = inPub(GeneSN))
```
```{r}
g <- ggplot(tb %>% arrange(publication), aes(x=logFC, y = -log2(adj.P.Val), text=GeneSN)) + 
  geom_point(aes(color = publication), size = 2) +
  scale_color_manual(values=c("grey", "red", "blue")) +
  theme(legend.title = element_blank(), 
        text = element_text(size = 14))
ggplotly(g, tooltip = "text")

```

### Most significantly different genes

```{r}
# Genes express high in AML
AML.hi.limma <- tb %>%
  filter(logFC > 1 & adj.P.Val < 0.05) %>%
  dplyr::select(probeset) %>%
  pull(1)

# Genes express high in ALL
ALL.hi.limma <- tb %>%
  filter(logFC < -1 & adj.P.Val < 0.05) %>%
  dplyr::select(probeset) %>%
  pull(1)

c(AML.hi.limma, ALL.hi.limma)
```
```{r}
# expression matrix of above 50 genes
require(gplots)
expr.heatmap <- train.log[c(AML.hi.limma, ALL.hi.limma),]
distCor <- function(x) as.dist(1-cor(x))
zClust <- function(x, scale="col", zlim=c(-3,3), method="average") {
  if (scale=="row") z <- t(scale(t(x)))
  if (scale=="col") z <- scale(x)
  z <- pmin(pmax(z, zlim[1]), zlim[2])
  hcl_row <- hclust(distCor(t(z)), method=method)
  hcl_col <- hclust(distCor(z), method=method)
  return(list(data=z, Rowv=as.dendrogram(hcl_row), Colv=as.dendrogram(hcl_col)))
}

z <- zClust(as.matrix(expr.heatmap), "row")

colrainbow <- rainbow(2)
heatmap.3(z$data, col=colorRampPalette(c("blue","white", "red"))(256), dendrogram="none", ColSideColors=as.matrix(colrainbow[as.factor(train.pheno$ALL.AML)]), Colv="none", Rowv="none", key=FALSE)

# s <- scale(t(expr.heatmap), scale=apply(t(expr.heatmap), 2, function(x) (max(x)-min(x))/3))
# 
# heatmap.3(t(s), col=colorRampPalette(c("blue","white", "red"))(256), dendrogram="none", ColSideColors=as.matrix(colrainbow[as.factor(train.pheno$ALL.AML)]), Colv="none", Rowv="none", key=FALSE)
```

## heatmap of 50 genes in the paper
```{r}
ps.AML.hi <- c("M55150", "X95735", "U50136", "M16038", "U82759", "M23197", "M84526", "Y12670", "M27891", "X17042", "Y00787", "M96326", "U46751", "M80254", "L08246", "M62762", "M28130", "M63138", "M57710", "M69043", "M81695", "X85116", "M19045", "M83652", "X04085")
ps.ALL.hi <- c("U22376","X59417","U05259","M92287","M31211","X74262","D26156","S50223","M31523","L47738","U32944","Z15115","X15949","X63469","M91432","U29175","Z69881","U20998","D38073","U26266","M31303","Y08612","U35451","M29696","M13792")

gs <- match(c(ps.AML.hi, ps.ALL.hi), sapply(rownames(train.log), function(x) strsplit(x, "_")[[1]][1]))
z <- zClust(as.matrix(train.log[gs, ]), "row")

colrainbow <- rainbow(2)

heatmap.3(z$data, col=colorRampPalette(c("blue","white", "red"))(256), dendrogram="none", Rowv="none", ColSideColors=as.matrix(colrainbow[as.factor(train.pheno$ALL.AML)]), Colv="none", key=FALSE)
```
```{r}
heatmap.3(train.log[grep(paste(c(ps.AML.hi, ps.ALL.hi), collapse="|"), rownames(train.log)), ], col=colorRampPalette(c("blue","white", "red"))(256), dendrogram="none", Rowv="none", ColSideColors=as.matrix(colrainbow[as.factor(train.pheno$ALL.AML)]), Colv="none", key=FALSE)
```
 
 
 unlogged data
 
```{r}
z <- zClust(as.matrix(train[grep(paste(c(ps.AML.hi, ps.ALL.hi), collapse="|"), rownames(train.log)), ]), "row")

colrainbow <- rainbow(2)

heatmap.3(z$data, col=colorRampPalette(c("blue","white", "red"))(256), dendrogram="none", Rowv="none", ColSideColors=as.matrix(colrainbow[as.factor(train.pheno$ALL.AML)]), Colv="none", key=FALSE)
```
 


## Construct predictor using 50 genes
```{r}
expr.train.50 <- as.data.frame(t(expr.heatmap))
expr.train.50$label <- as.character(train.pheno$ALL.AML)
# Create model with default paramters
control <- trainControl(method="repeatedcv", number=10, repeats=3)
seed <- 7
metric <- "Accuracy"
set.seed(seed)
mtry <- sqrt(ncol(expr.train.50))
tunegrid <- expand.grid(.mtry=mtry)
rf_default <- train(label~., data=expr.train.50, method="rf", metric=metric, tuneGrid=tunegrid, trControl=control, preProcess="scale")
print(rf_default)
```

## Testing 

```{r}
data("Golub_Test")
test <- exprs(Golub_Test)
# remove first 59 rows of controls
test <- test[60:nrow(test),]

test.log <- log2(pmax(test, 1))

test.pheno <- phenoData(Golub_Test)@data
```
```{r}
# check expression distribution
test.log.long <- as.data.frame(test.log) %>%
  rownames_to_column(var="Probeset") %>%
  gather(Sample, Expr, -Probeset)

ggplot(test.log.long, aes(x=Expr, group=Sample)) +
  stat_density(aes(color = Sample), geom="line", position="identity")
```


```{r}
# prediction

expr.test.50 <- as.data.frame(t(test.log[c(AML.hi.limma, ALL.hi.limma),]))
pred <- predict(rf_default, expr.test.50)

confusionMatrix(pred, reference = test.pheno$ALL.AML)
```

```{r}
# filter probeset with A call in more than 80% of samples
callFilter <- apply(call.train, 1, function(x) (sum(x == "A" )/length(x)) < 0.8)
train.log.callftd <- train.log[callFilter,]

for (i in 1:ncol(train.log.callftd)) {
  plotMA(train.log.callftd, array=i)
}
```

```{r}
fit.c <- lmFit(train.log.callftd, design)

fit2.c <- contrasts.fit(fit.c, cont.matrix)
fit2.c <- eBayes(fit2.c)
tb.c <- topTable(fit2.c, adjust="BH", number=nrow(train.log.ftd)) %>%
  rownames_to_column(var="probeset")
```

```{r}
# volcano plot and highlight the genes in paper




tb.c <- tb.c %>%
  rowwise() %>%
  mutate(GeneSN = strsplit(probeset, "_")[[1]][1]) %>%
  mutate(publication = inPub(GeneSN))
```
```{r}
g <- ggplot(tb.c %>% arrange(publication), aes(x=logFC, y = -log2(adj.P.Val), text=probeset)) + 
  geom_point(aes(color = publication), size = 0.8) +
  # geom_point(aes(color = publication), subset = .(publication != "None")) +
  scale_color_manual(values=c("grey", "red", "blue")) +
  theme(legend.title = element_blank())
ggplotly(g, tooltip = "text")
```

## No normalization 

```{r}
# remove low variance genes
sd.allgene <- apply(train, 1, sd)
summary(sd.allgene)
```
```{r}
fit.nonorm <- lmFit(train, design)
cont.matrix <- makeContrasts(AMLvsALL=AML-ALL, levels=design)
fit2.nonorm <- contrasts.fit(fit.nonorm, cont.matrix)
fit2.nonorm <- eBayes(fit2.nonorm)
tb.nonorm <- topTable(fit2.nonorm, adjust="BH", number=nrow(train.log.ftd)) %>%
  rownames_to_column(var="probeset")

tb.nonorm <- tb.nonorm %>%
  rowwise() %>%
  mutate(GeneSN = strsplit(probeset, "_")[[1]][1]) %>%
  mutate(publication = inPub(GeneSN))
```
```{r}
g <- ggplot(tb.nonorm %>% arrange(publication), aes(x=logFC, y = -log2(adj.P.Val), text=probeset)) + 
  geom_point(aes(color = publication), size = 0.8) +
  # geom_point(aes(color = publication), subset = .(publication != "None")) +
  scale_color_manual(values=c("grey", "red", "blue")) +
  theme(legend.title = element_blank())
ggplotly(g, tooltip = "text")
```

## Clustering

```{r}
cor <- cor(expr, method="spearman")
d <- as.dist(1-cor)

hc <- hclust(d, method="complete")

require(ggdendro)
ggdendrogram(hc, rotate = FALSE, size = 2)
```
```{r}
# K-means
km <- kmeans(scale(t(expr)), 2, nstart = 25)
```
## SOM
```{r}
library(kohonen)
sommap <- som(scale(t(expr)), grid = somgrid(6, 6, "hexagonal"))

## use hierarchical clustering to cluster the codebook vectors
groups<-2

som.hc <- cutree(hclust(dist(sommap$codes[[1]]), method="complete"), groups)

df.som <- data.frame(unit=sommap$unit.classif) %>%
  left_join(data.frame(unit=1:36, cluster=som.hc), by=c("unit" = "unit"))
#plot
plot(sommap, type="mapping", bgcol=rainbow(groups)[som.hc])

#cluster boundaries
add.cluster.boundaries(sommap, som.hc)
```
```{r}
den<-data.frame(cluster=paste0("hc_", cutree(hc,2)),Sample=factor(hc$labels, levels=hc$labels[hc$order]))

den <- den %>%
  mutate(ALL.AML=pheno$ALL.AML, kmeans=paste0("km_", km$cluster), som=paste0("SOM_", df.som$cluster))

p1<-ggdendrogram(hc, rotate=FALSE)


p2<-ggplot(den,aes(Sample,y=4,fill=factor(cluster)))+
  geom_tile()+
  geom_tile(aes(Sample, y=1, fill=factor(ALL.AML))) +
  geom_tile(aes(Sample, y=2, fill=factor(kmeans))) +
  geom_tile(aes(Sample, y=3, fill=factor(som))) +
  scale_y_continuous(expand=c(0,0))+
  scale_fill_manual(values=rainbow(8)[c(1, 5, 2, 6, 3, 7, 4, 8)])+
  theme(axis.title=element_blank(),
        axis.ticks=element_blank(),
        axis.text=element_blank()
        )
```
```{r}
lab <- cutree(hc, 2)
table(pheno$ALL.AML, lab)
```
## Venn diagram of Mark's gene
```{r}
m.genes <- read.table(file="markGenes.txt", header=T, stringsAsFactors = F)

require(hu6800.db)
x <- hu6800ENTREZID
mapped_probes <- mappedkeys(x)
xx <- toTable(x[mapped_probes])

geneDesc <- unlist(xx[m.genes$Probeset])

library(org.Hs.eg.db)
e2s = toTable(org.Hs.egSYMBOL)

m.genes <- m.genes %>%
  left_join(xx, by=c("Probeset" = "probe_id")) %>%
  left_join(e2s, by = "gene_id")

write.table(m.genes, file="markGenes.txt", sep="\t", row.names = F, quote=F)
```
```{r}
require(VennDiagram)

venn.diagram(list(Ours = m.genes$gSesNum[26:50], Theirs = ps.AML.hi), filename = "vennAML.tiff", fill=rainbow(4)[1:2])
venn.diagram(list(Ours = m.genes$gSesNum[1:25], Theirs = ps.ALL.hi), filename = "vennALL.tiff", fill=rainbow(4)[3:4])
```

## rescaling factors
```{r}
getRFactor <- function(a, b) {
  eSub <- dta.all[, c(a, b)]
  cSub <- call.all[, c(a, b)]
  pRows <- which(apply(cSub, 1, function(x) all(x == "P")))
  eSub <- eSub[pRows, ]
  colnames(eSub) <- c("ref", "test")
  fit <- lm(test~ref, data=eSub)
  rFactor <- 1/(fit$coefficients[2])
  return(rFactor)
}
  
```

```{r}
dta.all <- cbind(dta.train, dta.test)
call.all <- cbind(call.train, call.test)

rFactors <- NULL
for (i in 2: ncol(dta.all)) {
  rFactors <- c(rFactors, getRFactor(1, i))
}

summary(rFactors)
```
```{r}
tb.rf <- read.table("table_ALL_AML_rfactors.txt", sep="", header=F, skip=2, col.names = c("Sample", "rFactor"))
tb.rf$Source <- "Golub"

newR <- data.frame(Sample=as.numeric(colnames(dta.all)), rFactor=c(1, rFactors), Source="Ours")

tb.rf <- rbind(tb.rf, newR)

ggplot(tb.rf, aes(x=factor(Sample), y = rFactor)) + 
  geom_col(aes(fill=Source)) +
  facet_grid(Source ~ .) + 
  xlab("Sample") +
  theme(axis.text.x=element_text(angle=90, vjust=0.5))
```



